docs/12-entity-component-system.mdx

```diff
@@ -0,0 +1,150 @@
+- [[System]] - Implementation of behavior given a set of components on an entity
+
+Many of the features built in Excalibur are built using this, and users of Excalibur can make their own systems and/or replace
+existing excalibur systems.
+
+## Excalibur ECS Implementation
+
+In addition to the familiar [[Entity]], [[Component]], and [[System]], Excalibur also implements an ECS [[World]] where all of these
+pieces live and can interact together. Each [[Scene]] in Excalibur conatins an ECS world, they are not shared between scenes.
+
+Inside an ECS [[World]] there is an [[EntityManager]], [[QueryManager]], and [[SystemManager]] that handle all the details.
+
+- [[EntityManager]] - Manages all the entities known by the world, it is an [[Observer]] of entity component changes over time.
+- [[QueryManager]] - Manages all the entity queries, a [[Query]] allows a search of all known entities by a set of component types.
+- [[SystemManager]] - Manages all the systems known by the world, it updates a [[System]] in priority order (low to high) and gives them a set of entities that match types.
+
+Excalibur [[Actor]]'s are in fact [[Entity]]'s in this ECS implementation. Actors come "pre-installed" with a set of components and features.
+
+## Built in Excalibur Components
+
+### [[TagComponent]]
+
+Tag components are useful when you want to flag an entity with something, for example "offscreen". A good way to think about it is to use a flag for a non-default state.
+
+```ts
+const entity = new ex.Entity()
+entity.addComponent(new ex.TagComponent('offscreen'))
+
+console.log(entity.tags) // ['offscreen']
+```
+
+### [[TransformComponent]]
+
+The Excalibur TransformComponent component primarily keeps track of the [[TransformComponent.z|z-index]] [[TransformComponent.pos|position]], [[TransformComponent.rotation|rotation]], and [[TransformComponent.scale|scale]] of an entity.
+
+Additionally it keeps track of the [[TransformComponent.coordPlane|coordinate plane]] which is whether its draw sticky to the [[CoordPlane.Screen]] or part of the [[CoordPlane.World]].
+
+### [[CanvasDrawComponent]]
+
+The Excalibur CanvasDrawComponent component tells [[CanvasDrawSystem]] how to draw to a HTML 2D Canvas.
+
+```ts
+const entity = new ex.Entity()
+entity.addComponent(new TransformComponent())
+entity.addComponent(
+  new CanvasDrawComponent((ctx, delta) => {
+    // draw stuff
+    ctx.fillStyle = 'black'
+    ctx.fillRect(0, 0, 200, 200)
+  })
+)
+```
+
+## Built in Excalibur Systems
+
+### [[CanvasDrawSystem]]
+
+The Excalibur CanvasDrawSystem takes any entity with a [[TransformComponent]] and a [[CanvasDrawComponent]] and draws it to the canvas using the 2D canvas api.
+
+## Implementing your own Components & Systems
+
+To build your own component, extend the Excalibur [[Component]] abstract class and pick a unique type name (duplicate type names will cause problems).
```

Comment:
Do we have duplicate type name detection in the ECS implementation?

Reply:
This is tricky, I haven't come up with a way yet. It might be possible to also use the runtime typeof a component to disambiguate. But more research is required